XP-456: replace CLI argument with a config file

little-dude opened the pull request:
### Summary

replace the CLI arguments with a config file

### Reference:

https://xainag.atlassian.net/browse/XP-456
https://xainag.atlassian.net/browse/XP-512

### Rationale

The CLI is getting complex so it is worth loading the configuration
from a file instead.

### Implementation details

#### TOML

We decided to use TOML for the following reasons:

  - it is human friendly, ie easy to read and write
  - our configuration has a pretty flat structure which makes TOML
    quite adapted
  - it is well specified and has lots of implementation
  - it is well known

The other options we considered:

  - INI: it is quite frequent in the Python ecosystem to use INI for
    config files, and the standard library even provides support for
    this. However, INI is not as powerful as TOML and does not have a
    specification
  - JSON: it is very popular but is not human friendly. For instance,
    it does not support comments, is very verbose, and breaks
    easily (if a trailing comma is forgotten at the end of a list for
    instance)
  - YAML: another popular choice, but is in my opinion more complex
    than TOML.


#### Validation

We use the third-party `schema` library to validate the
configuration. It provides a convenient way to:

- declare a schema to validate our config
- leverage third-party libraries to validate some inputs (we use the
  `idna` library to validate hostnames)
- define our own validators
- transform data after it has been validated: this can be useful to
  turn a relative path into an absolute one for example
- provide user friendly error message when the configuration is
  invalid

#### The `Config` class

By default, the `schema` library returns a dictionary containing a
valid configuration, but that is not convenient to manipulate in
Python. Therefore, we dynamically create a `Config` class from the
configuration schema, and instantiate a `Config` object from the data
returned by the `schema` validator.

#### Package re-organization

We moved the command line and config file logic into its own `config`
sub-package, and moved the former `xain_fl.cli.main` entrypoint into
the `xain_fl.__main__` module.

#### Docker infrastructure

- Cache the xain_fl dependencies. This considerably reduces
  "edit->build-> debug" cycle, since installing the dependencies takes
  about 30 minutes.
- Move all the docker related files into the `docker/` directory

#### Limitations and future work

1. The documentation generated for the `ServerConfig`,
   `AiConfig` and `StorageConfig`  classes is wrong. Each attribute is
   documented as "Alias for field number X". This can be fixed by
   having `create_class_from_schema()` setting the `__doc__` attribute
   for each attribute. However, we won't be able to automatically
   document the type of each attribute.

2. When the configuration contains an invalid value, the error message
   we generate does not contain the invalid value in question. I think
   it is possible to enable this in the future but haven't really
   looked into it.

---

### Reviewer checklist

*Reviewer agreement:*

* Reviewers assign themselves at the start of the review.
* Reviewers do **not** commit or merge the merge request.
* Reviewers have to check and mark items in the checklist.

**Merge request checklist**

- [x] Conforms to the merge request title naming `XP-XXX <a description in imperative form>`.
- [x] Each commit conforms to the naming convention `XP-XXX <a description in imperative form>`.
- [x] Linked the ticket in the merge request title or the references section.
- [x] Added an informative merge request summary.

**Code checklist**

- [x] Conforms to the branch naming `XP-XXX-<a_small_stub>`.
- [x] Passed scope checks.
- [x] Added or updated tests if needed.
- [x] Added or updated code documentation if needed.
- [x] Conforms to Google docstring style.
- [x] Conforms to XAIN structlog style.

Robert-Steiner commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,358 @@
+    {
+        "initial_weights": Use(str, error=error("ai.initial_weights", "a valid path"),),
+        "rounds": strictly_positive_integer("ai.rounds"),
+        "epochs": strictly_positive_integer("ai.epochs"),
```
Here we need to change the validation, so that we accept `0` as well.
See https://github.com/xainag/xain-fl/pull/227

little-dude replied:
Thank you, done!

Robert-Steiner commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,358 @@
+
+
+def strictly_positive_integer(
+    key: str, expected_value: str = "a strictly positive_integer"
```
```suggestion
    key: str, expected_value: str = "a strictly positive integer"
```

Robert-Steiner commented on docker/dev/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Cache the dependencies
+COPY setup.py .
+RUN mkdir xain_fl && \
+    printf '__version__ = "0.2.0"\n__short_version__ = "0.2"' > xain_fl/__version__.py && \
```
Do we need to update the version values when we release a new package or are they just dummy values for the `setup.py`?

little-dude replied:
They're just dummy values for the `setup.py` indeed.

dkravetz commented:
The scope of this PR is pretty large. Please limit it to the ticket that it refers to

dkravetz commented on docker/dev/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    touch README.md && \
+    python setup.py egg_info && \
+    cat *.egg-info/requires.txt | grep -v '^\[' | sort | uniq | pip install -r /dev/stdin
+
```
Feel free to copy setup.py, cat/grep the version of gRPC, pip install it, and then install the full set of dependencies. They will cache the built package of grpc. Otherwise this is too confusing, specially for an open source project, with a lot of magic constants that make this Dockefile harder to read than necessary

little-dude replied:
That's what I did initially but for some reason gRPC was getting re-installed when I installed `xain-fl` anyway, and I was not able to understand why. When installing _all_ the dependencies at once, things are not getting re-installed. I can add a more informative comment about what this does though.

little-dude replied:
I added a comment to explain what's going on and why we're doing it that way.

dkravetz commented on docker/dev/docker-compose.yml:
```diff
@@ -48,20 +48,14 @@ services:
-      context: .
-      dockerfile: Dockerfile.dev
-    command: sh -c "coordinator -f test_array.npy --storage-endpoint http://minio-dev:9000 --storage-key-id $${MINIO_ACCESS_KEY} --storage-secret-access-key $${MINIO_SECRET_KEY} --storage-bucket xain-fl-aggregated-weights"
+    image: xain-fl-dev
```
The idea of the "dev" image is to build and run locally, not to pull from a registry. By removing the `build` directive, an image by the name of "xain-fl-dev" will try to be pulled from docker hub

little-dude replied:
Ah yes, I did that temporarily because I wanted to docker-compose to reuse the local image I had already built with `docker build`. I forgot to undo.

little-dude replied:
I reverted this change.

atymoshchuk commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,358 @@
+    """
+
+    def __init__(self, err, *args, **kwargs):
+        super().__init__(f"Invalid configuration:\n{err.code}", *args, **kwargs)
```
Please use standard exception message, and then structlog to log a message (our best practices).

little-dude replied:
I'm not sure what you mean. Do you have an example somewhere else to show me?

little-dude replied:
Also I think we don't want to print a stacktrace here. These errors are legit (as in "not an un-expected error in our code") at we should probably print a nice error message to the users instead.

atymoshchuk replied:
Sure, we don’t want to print here anything.
The way how we handle our exceptions (best practices in the team):
1. create a custom exception,
2. catch this exception where it can occur,
3. handle exception and also log useful information (using structlog).

Basically message is not needed in custom exception and we don’t print anything, we log errors.

little-dude replied:
Thanks. I pushed some changes that address this I believe :)

little-dude commented on Dockerfile.dev:
```diff
@@ -1,12 +0,0 @@
-FROM python:3.6-alpine
```
I moved the docker stuff in the docker directory because I added two files to put in the image, and things were getting messy.

little-dude replied:
I can create a ticket for that if you want?

little-dude commented on Dockerfile.dev:
```diff
@@ -1,12 +0,0 @@
-FROM python:3.6-alpine
-
```
Resolving this conversation because I think it's a duplicate of the one above.

little-dude commented:
> Please create follow-up tickets to make changes to docker files and any other changes. Would be great to have only changes about configuration in this pr and everything else in other dedicated prs.

Without the docker changes, this PR will break the docker images, is that ok?

rsaffi commented:
Although I'm not listed as reviewer, by @dkravetz's request I went through the changes and here are my 2 cents:

- Code-wise it looks good
- The scope of the changes go beyond the scope of the ticket that the PR references (XP-456), so ideally I'd break the sets of changes into two different PRs: one that is managing the replacement of CLI arguments for the config file and another for XP-512, for the caching of dependencies in docker.

That way, whenever commits are squashed together and merged, we have changes that are pertinent to only one ticket.

little-dude commented:
I reverted moving around the docker files @dkravetz @rsaffi.
However I'd still like to make the caching changes and having a working `Dockerfile.dev` in this PR because:

- I need this to manually test the storage and some people may also be relying on the docker image
- with the shuffling around reverted, the changes are pretty small and easy to review
- I making the minimal required changes to keep the docker environment working when you make changes to the code can be considered as being in scope of a given ticket
- the work is already done and splitting it out means spending more time on this

little-dude commented:
@dkravetz 

> The scope of this PR is pretty large. Please limit it to the ticket that it refers to

Understood, I reverted my most invasive changes. I just updated the `Dockerfile.dev` and `docker-compose-dev.yml` to work with the new code. I hope that's OK.

little-dude commented on tests/test_config.py:
```diff
@@ -0,0 +1,173 @@
+# pylint: disable=missing-docstring,redefined-outer-name
```
Ah I should actually remove the `missing-docstring` one.
But the `redefined-outer-name` should actually be set for the whole `tests` subpackage, because it's how fixtures work in pytest: each test function/method's argument shadows an outer name.

little-dude replied:
I removed the `missing-docstring` lint, but as said above I'd like to keep the `redefined-outer-name` one, if that's ok for you.

janpetschexain replied:
ok, thanks for adding a comment in code explaining the globally disabled warning!

little-dude commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,355 @@
+    return f"Invalid `{key}`: value must be {description}"
+
+
+def strictly_positive_integer(
```
That's funny. In french _strictly positive_ unambiguously means `> 0` while _positive_ means `>= 0`. In english, _strictly positive_ [seems to be used as well](https://en.wikipedia.org/wiki/Strictly_positive_measure) but it seems that _positive_ [can also mean `> 0`](https://math.stackexchange.com/questions/18464/is-positive-the-same-as-non-negative). Maybe let's ask @finiteprods what the preferred terminology is?

finiteprods replied:
I think you both sum up the situation pretty well! I wish we could say "natural number" and put this to rest but even there, there's sometimes (a similar) disagreement ;-) In my experience, "positive" _usually_ means `> 0` since we tend to say "non-negative" for `>= 0`. If you feel passionate about this, you could keep `strictly_positive_integer` but rename the other one to `nonnegative_integer`, otherwise just go with @janpetschexain 's suggestion.

little-dude replied:
Thanks @finiteprods! I applied your suggestion @janpetschexain :)

little-dude commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,355 @@
+)
+
+
+# pylint: disable=protected-access
```
Thanks, I moved it to the appropriate line.

Robert-Steiner commented on README.md:
````diff
@@ -99,6 +99,19 @@ entrypoint:
+  xain-fl-dev \
+  coordinator --config /custom_config.toml
+```
+
````
Just one small additional thing:

Since we don't have the `cli.py` file anymore, we need to remove or update the `Running the Coordinator locally`  section in the `README.md`

Robert-Steiner replied:
When I try to run this example I get the following error:
```
Traceback (most recent call last):
  File "/usr/local/bin/coordinator", line 11, in <module>
    load_entry_point('xain-fl', 'console_scripts', 'coordinator')()
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 490, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2853, in load_entry_point
    return ep.load()
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2444, in load
    return self.resolve()
  File "/usr/local/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2450, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/app/xain_fl/__main__.py", line 41, in <module>
    main()
  File "/app/xain_fl/__main__.py", line 23, in main
    config = Config.load(args.config)
  File "/app/xain_fl/config/schema.py", line 345, in load
    with open(path, "r") as f:
IsADirectoryError: [Errno 21] Is a directory: '/custom_config.toml'
```
Can we maybe just use the default config?

```
coordinator --config docker/dev/xain-fl.toml 
```

little-dude replied:
This section explains how to pass a custom config to the container by mounting a config file and a `.npy` file in the container:

```shell
  -v $(pwd)/custom_config.toml:/custom_config.toml \
  -v $(pwd)/custom_initial_weights.npy:/custom_initial_weights.npy \
```

In this command, if `$(pwd)/custom_config.toml` and `$(pwd)/custom_initial_weights.npy` don't exist on your host, docker creates empty directories in the container. This is why you get this error message.

Try this out:

```
cp docker/dev/initial_weights.npy custom_initial_weights.npy
cp docker/dev/xain-fl.toml custom_config.toml
sed -i  's#^initial_weights.*#initial_weights = /custom_initial_weights.npy#g' custom_config.toml
docker run \
  -v $(pwd)/custom_config.toml:/custom_config.toml \
  -v $(pwd)/custom_initial_weights.npy:/custom_initial_weights.npy \
  -v $(pwd):/app \
  -v '/app/xain_fl.egg-info' \
  xain-fl-dev \
  coordinator --config /custom_config.toml
```

> Can we maybe just use the default config?

The previous command in the README does that already, but I want to document how to pass a custom config as well, because I'm pretty sure someone will need it, and it's not that obvious.

little-dude replied:
I pushed a few changes. First I made the README clearer about what this command does.

Also we now log proper error and exit if we cannot open the config file or read it.

For instance if the file is actually a directory (your case):
```
❯ sudo rm -r custom_config.toml custom_initial_weights.npy
❯ docker run \
  -v $(pwd)/custom_config.toml:/custom_config.toml \
  -v $(pwd)/custom_initial_weights.npy:/custom_initial_weights.npy \
  -v $(pwd):/app \
  -v '/app/xain_fl.egg-info' \
  xain-fl-dev \
  coordinator --config /custom_config.toml

{
  "event": "Invalid config: /custom_config.toml is a directory",
  "level": "error"
}
```

If the file contains invalid TOML:
```
❯ cp docker/dev/initial_weights.npy custom_initial_weights.npy                                                                          
cp docker/dev/xain-fl.toml custom_config.toml                                                                                           
sed -i  's#^initial_weights.*#initial_weights = /custom_initial_weights.npy#g' custom_config.toml                                       
docker run \
  -v $(pwd)/custom_config.toml:/custom_config.toml \
  -v $(pwd)/custom_initial_weights.npy:/custom_initial_weights.npy \ 
  -v $(pwd):/app \
  -v '/app/xain_fl.egg-info' \
  xain-fl-dev \
  coordinator --config /custom_config.toml


{
  "event": "Invalid config: failed to decode /custom_config.toml: This float doesn't have a leading digit (line 9 column 1 char 224)",
  "level": "error"
}
```

rsaffi commented on Dockerfile.dev:
```diff
@@ -3,10 +3,28 @@ FROM python:3.6-alpine
 COPY setup.py .
-COPY xain_fl xain_fl/
+RUN mkdir xain_fl && \
+    printf '__version__ = "0.2.0"\n__short_version__ = "0.2"' > xain_fl/__version__.py && \
```
It shouldn't make much of a difference, but at the moment of this writing `v0.3.0` has already been released for `xain-fl`: https://github.com/xainag/xain-fl/pull/230

little-dude replied:
These are dummy numbers anyway. I should perhaps make it clearer by making it just "0".

little-dude replied:
Done. Changed it to `0`.

rsaffi commented on docker/dev/xain-fl.toml:
```diff
@@ -0,0 +1,29 @@
+
+[storage]
+# URL to the storage service to use
+endpoint = "http://minio-dev:900"
```
This should be port 9000, right?

little-dude replied:
geez thanks, yes of course

little-dude replied:
fixed.

rsaffi commented on example-config.toml:
```diff
@@ -0,0 +1,29 @@
+
+[storage]
+# URL to the storage service to use
+endpoint = "http://minio-dev:900"
```
Same, port should be 9000.

finiteprods commented on Dockerfile.dev:
```diff
@@ -3,10 +3,28 @@ FROM python:3.6-alpine
+#
+# However, we cannot just pre-install a few packages, because we don't
+# know exactly which version `pip` will pick for them. Instead, we
+# give pip all the package's dependencies, and let it resolve and
```
```suggestion
# give `pip` all the package's dependencies, and let it resolve and
```

finiteprods commented on README.md:
```diff
@@ -68,16 +68,13 @@ $ make show
 ### Running the Coordinator locally
 
-To run the Coordinator on your local machine, use the command:
+To run the Coordinator on your local machine you can use the
```
```suggestion
To run the Coordinator on your local machine, you can use the
```

finiteprods commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+#
+# Thus, although we use schemas to validate configurations (AI_SCHEMA,
+# SERVER_SCHEMA, and STORAGE_SCHEMA), we don't expose them. Instead,
+# we use them to dynamically generate classes which attributes are the
```
```suggestion
# we use them to dynamically generate classes where attributes are the
```

finiteprods commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+# Thus, although we use schemas to validate configurations (AI_SCHEMA,
+# SERVER_SCHEMA, and STORAGE_SCHEMA), we don't expose them. Instead,
+# we use them to dynamically generate classes which attributes are the
+# schemas keys: AiConfig, ServerConfig, and StorageConfig. This
```
```suggestion
# schema keys: AiConfig, ServerConfig, and StorageConfig. This
```

finiteprods commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+
+
+def non_negative_integer(key: str, expected_value: str = "a positive integer") -> Schema:
+    """Return a positive integer validator for the given configuration
```
```suggestion
    """Return a non-negative integer validator for the given configuration
```

finiteprods commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+
+    Returns:
+
+        A new class which attributes are the given schema's keys
```
```suggestion
        A new class where attributes are the given schema's keys
```

finiteprods commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+       # AWS access key ID to use to authenticate to the storage service
+       access_key_id = "my-key-id"
+
+    This config file can be loaded and used as follow:
```
```suggestion
    This config file can be loaded and used as follows:
```

finiteprods commented on tests/test_config.py:
```diff
@@ -0,0 +1,187 @@
+
+# Adapted from unittest's assertRaises
+class AssertInvalid:
+    """A context manager that check that an `xainfl.config.InvalidConfig`
```
```suggestion
    """A context manager that checks that an `xainfl.config.InvalidConfig`
```

finiteprods commented on tests/test_config.py:
```diff
@@ -0,0 +1,187 @@
+        if exc_type is None:
+            raise Exception("Did not get an exception")
+        if not isinstance(exc_value, InvalidConfig):
+            # let this un-expected exception be re-raised
```
```suggestion
            # let this unexpected exception be re-raised
```

little-dude commented:
Thanks for the review @finiteprods, I applied the changes you suggested.

atymoshchuk commented on .gitignore:
```diff
@@ -129,3 +129,8 @@ dmypy.json
+# In the README, we document how to use custom config. Let's ignore
+# the files mentioned in that section
+custom_config.toml
+custom_initial_weights.npy
```
New line is missing here.

atymoshchuk commented on Dockerfile.dev:
```diff
@@ -3,10 +3,28 @@ FROM python:3.6-alpine
 COPY setup.py .
-COPY xain_fl xain_fl/
+RUN mkdir xain_fl && \
+    printf '__version__ = "0"\n__short_version__ = "0"' > xain_fl/__version__.py && \
```
Maybe I missed something in code review from before, but why do we need this dummy version here?

little-dude replied:
Just so that `setup.py` can run. `setup.py` opens this file.

atymoshchuk commented on README.md:
```diff
@@ -68,16 +68,13 @@ $ make show
-$ python xain_fl/cli.py --help
+coordinator --config example-config.toml
+# or
+python xain_fl --config example-config.toml
```
It's a bit confusing to have two different commands to run app with no explanations which one is better and what's the difference between them. Will it be possible to add one line of explanations?

little-dude replied:
As far as I know these two commands are pretty much equivalent, except that the second one doesn't require `xain_fl` to be installed. I'll just remove the second one maybe?

atymoshchuk replied:
We can leave both of them, the one for development and one for running coordinator locally, would be good to have short explanation of the difference.

atymoshchuk commented on docker/dev/xain-fl.toml:
```diff
@@ -0,0 +1,29 @@
+[server]
```
Configuration files do not belong to docker, would be great to move them to dedicated folder with configurations.
On the side note: as we are not going to have any other configurations except dev and example published in open source, make sense not to create a dedicated folder to the only configuration we have, what do you think?

little-dude replied:
Some parts of the config are specific to docker image we build. For example the path to the file that contains the initial weights is `/initial_weights.npy` in docker, but will most likely differ from one environment to another. That's why I put them in the `docker/` directory. I'm open to moving them, let me know if you prefer.

little-dude replied:
I moved the config file to the top of the repo and remove `docker/dev`

atymoshchuk commented on docker/dev/xain-fl.toml:
```diff
@@ -0,0 +1,29 @@
+# AWS access key ID to use to authenticate to the storage service
+access_key_id = "minio"
+# AWS secret access to use to authenticate to the storage service
+secret_access_key = "minio123"
```
new line is missing

atymoshchuk commented on xain_fl/__main__.py:
```diff
@@ -0,0 +1,41 @@
+    try:
+        config = Config.load(args.config)
+    except InvalidConfig as err:
+        logger.error(f"Invalid config: {err}")
```
As we use structlog, would be nice to have structure here, and one of the best practices is not to use f-strings in logs: [More info about issue with f-strings](https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs/54368086)
Structlog log entry looks like:
```suggestion
        logger.error("Invalid config", error=repr(err))
```
repr will give us type of the error, we can also log any other specific information as a key-value pair.

atymoshchuk commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+        description (str): description of the expected type of value
+            for this configuration item
+    """
+    return f"Invalid `{key}`: value must be {description}"
```
Please update this function after updating structlog message.

little-dude replied:
@atymoshchuk sorry, I'm not sure what you mean here.

atymoshchuk commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+        try:
+            valid_config = CONFIG_SCHEMA.validate(dictionary)
+        except SchemaError as err:
+            raise InvalidConfig(err.code)
```
in this case we are raising another error while handling the one we got, which is not our intention i beliebe.
More information how to handle exceptions can be found here: [Talk about exceptions by Mario Corchero](https://www.youtube.com/watch?v=V2fGAv2R5j8)
```suggestion
            raise InvalidConfig(err.code) as err
```

atymoshchuk commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+            with open(path, "r") as f:
+                raw_config = toml.load(f)
+        except IsADirectoryError:
+            raise InvalidConfig(f"{path} is a directory")
```
same here:
```suggestion
    except IsADirectoryError as err:
            raise InvalidConfig(f"{path} is a directory") as err
```

atymoshchuk commented on xain_fl/config/schema.py:
```diff
@@ -0,0 +1,389 @@
+        except FileNotFoundError:
+            raise InvalidConfig(f"{path} not found")
+        except PermissionError:
+            raise InvalidConfig(f"failed to read {path}: insufficient permissions")
```
in case if we would like to have some information in readable form, would be nice to log the error here and have more generic exception message.

little-dude replied:
I believe your referring to [my earlier comment](https://github.com/xainag/xain-fl/pull/221#discussion_r368863619)? If so I ended up changing my mind about this actually :p Logging these exceptions when we catch them seems to print clear enough logs. Here is an example: https://github.com/xainag/xain-fl/pull/221#discussion_r369185286

What do you think?

little-dude commented:
This conflicts with https://github.com/xainag/xain-fl/pull/237 but I'll rebase after the final review.

dkravetz commented on docker-compose-dev.yml:
```diff
@@ -48,19 +48,15 @@ services:
+      - ${PWD}/xain_fl:/app/xain_fl
       - ${PWD}/setup.py:/app/setup.py
       - ${PWD}/README.md:/app/README.md
+      - ${PWD}/configs/development.toml:/xain-fl.toml
```
The path should be updated to `/app/xain-fl.toml` (same as in the Dockerfile.dev)
